pynumaflow/function/_dtypes.py

```diff
@@ -13,32 +13,65 @@
 Ms = TypeVar("Ms", bound="Messages")
 MT = TypeVar("MT", bound="MessageT")
 MTs = TypeVar("MTs", bound="MessageTs")
+NPC = TypeVar("NPC", bound="NoPublicConstructor")
+
+
+class NoPublicConstructor(type):
+    """Metaclass that ensures a private constructor
+
+    If a class uses this metaclass like this:
+
+        class SomeClass(metaclass=NoPublicConstructor):
+            pass
+
+    If you try to instantiate your class using (`SomeClass()`), a `TypeError` will be thrown.
+    """
+
+    def __call__(cls, *args, **kwargs):
+        raise NoPublicConstructorError(
+            "public constructor is not supported, please use class methods to create the object."
+        )
+
+    def _create(cls: Type[NPC], *args: Any, **kwargs: Any) -> NPC:
+        return super().__call__(*args, **kwargs)
+
+
+class NoPublicConstructorError(TypeError):
```

Comment:
Nice, let's move this to a new file. Name it `exceptions.py` inside `pynumaflow/`

Reply:
+1. I also moved the `NoPublicConstructor` to `types.py` inside `pynumaflow/`, thinking it could be shared across the entire project.